HADOOP-16721. Improve S3A rename resilience

steveloughran opened the pull request:
S3A Rename to

1. switch from verifying dest parent path is present and a directory to simply verifying that it isn't a file.
2. avoids race condition where thread 1 has deleted the last subdir/file but not yet recreated a parent marker
3. don't downgrade FileNotFoundException to 'false' if source doesn't exist
4. raise FileAlreadyExistsException if dest path exists or parent is a file.

If thread/process 1 deleted the subdir `dest/subdir1` one and there
are no sibling subdirectories, then then the dir `dest` would not
exist until `maybeCreateFakeParentDirectory()` had performed a
`LIST` and, if needed, a `PUT` of a marker.
This creates a window where thread/process 2, trying to rename `staging/subdir2`
into `dest` could fail "parent does not exist".

And guess what:
1. Hive still thinks renaming directories from staging to the production dataset is a good idea.
1. It can spawn many threads to do this parallel (maybe even to compensate for the slow rename())
performance of S3.
1. On a sufficiently overloaded system (lots of threads already doing the parallel filenames)
the window of parent-dir-not-found can last long enough for things to fail.
   
Prior to S3 being consistent this wouldn't have been an issue
* rename()'s need to list everything underneath the source dir wasn't safe to use with delayed list consistency
* so S3Guard was effectively mandatory.
* and delayed LIST consistency may even have reduced the window size.

The fix: go from verifying parent dir exists to simply making sure that it isn't a file
is a weakening of the requirement "parent dir must exist" -but file:// already doesn't
require that. 

Consequences
1. You can rename under a path which doesn't exist. file:// lets you do that already; contract tests have a switch for it.
   (FS contract test has switch for this)
2. Exceptions get raised when false would be returned and hive lose details of what went wrong. This
   isn't consistent with HDFS, but is with other stores (FS contract test has switch for this)
3. You can rename a few layers under a file, but then `create()` lets you do that already,
  and nobody has noticed in production.
   
If directory marker retention was enabled there's more likelihood that an empty dir marker existed -if it did then the race condition wouldn't exist. But as there are no guarantees that the marker will be there, that's not safe to rely on.

steveloughran committed:
HADOOP-16721. s3a rename to support

fs.s3a.rename.raises.exceptions: raise exceptions on rename failures
fs.s3a.rename.reduced.probes: don't look for parent dir (LIST), just verify
 it isn't a file.

The reduced probe not only saves money, it avoids race conditions
where one thread deleting a subdir can cause LIST <parent> to fail before
a dir marker is recreated.

Note:

* file:// rename() creates parent dirs, so this isn't too dangerous.
* tests will switch modes.

We could always just do the HEAD; topic for discussion. This patch: optional

Change-Id: Ic0f8a410b45fef14ff522cb5aa1ae2bc19c8eeee

steveloughran commented:
Testing with  -Dparallel-tests -DtestsThreadCount=8 -Dmarkers=keep -Dscale

If the store is set to raise exceptions, a lot of tests which expect rename(bad options) to return false now get exceptions. One of the contract tests would downgrade if the raised exception was FileAlreadyExistsException and the contract xml said that was ok. I'm reluctant to go with a bigger patch.

this PR is so that Hive and friends can get better reporting on errors, rather than have them lost. It will be optional

steveloughran committed:
HADOOP-16721. S3A rename resilience

Check for parent dir is now always !file, not isDirectory; avoids
race conditions with parallel deletes of subdirectories under destination

Also switches connector to raise exceptions on conditions
which don't trigger exceptions on HDFS (just return false

* source file missing
* destination existing

Some other connectors already report these as failures. With this
change S3 moves away from HDFS behaviour, but towards one with better
error reporting, and which we know is handled in existing applications.

Documentated the changes in filesystem.md and troubleshooting s3a

Change-Id: I5daf54636e63c19f273c86519d5eeb3cbeffeb49

steveloughran committed:
HADOOP-16721 option for s3a rename to raise exceptions, instead of returning false.

about to revert.

Change-Id: I0d91330e2a59e8c6863ebeedce17573e69e2bef5

steveloughran committed:
HADOOP-16721 rename resilience

revert option to raise exceptions instead of returning false; explicitly
raising FileNotFoundException/FileAlreadyExistsException is better.

Change-Id: Ica78e0abf5a1444227e4843288679a2d3441dd70

steveloughran committed:
HADOOP-16721. Add integration test for the race condition

Change-Id: Ic8787cb6f87d4b5cda23c6ced1e73922b665e797

steveloughran commented:
Testing: s3 london with markers==keep and delete, s3guard on and off.

* change in error reporting of s3a FS matched with relevant changes in s3a.xml for contract tests.
* skip tests verifying that you can't rename 2+ levels under a file.
* Failures related to endpoints of common-crawl and  ITestAssumeRole.testAssumeRoleBadInnerAuth: known and fixed in #2675.

iwasakims commented on hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/troubleshooting_s3a.md:
```diff
@@ -1126,6 +1126,26 @@ We also recommend using applications/application
+* HEAD then LIST of destination path
+* File-by-file copy of source objects to destination.
+  Parallelized, with page listings of directory objects and issuing of DELETE requests.
+* Post-delete recreation of destination parent directory marker, if needed.  
```
recreation of **source** parent directory marker?

steveloughran replied:
oops. fixed

iwasakims commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/impl/ITestlRenameDeleteRace.java:
```diff
@@ -0,0 +1,200 @@
+ * This test suite recreates the failure using semaphores to guarantee the failure
+ * condition is encountered -then verifies that the rename operation is successful.
+ */
+public class ITestlRenameDeleteRace extends AbstractS3ATestBase {
```
Is the class name intended to be ITestRenameDeleteRace?

steveloughran replied:
yeah. well spotted

steveloughran committed:
HADOOP-16721. review comments

modified:   src/site/markdown/tools/hadoop-aws/troubleshooting_s3a.md
modified:   src/test/java/org/apache/hadoop/fs/s3a/S3ATestUtils.java
renamed:    src/test/java/org/apache/hadoop/fs/s3a/impl/ITestlRenameDeleteRace.java -> src/test/java/org/apache/hadoop/fs/s3a/impl/ITestRenameDeleteRace.java

moved the "disable s3guard" test setup into S3ATestUtils; new suites are going
to be adopting it until we pull out S3Guard completely.

Change-Id: I0303aec2f6e4c98685da62e9f31c7e44a1e67d41

iwasakims commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/impl/ITestlRenameDeleteRace.java:
```diff
@@ -0,0 +1,200 @@
+    /**
+     * Semaphore to acquire before the marker can be listed/created.
+     */
+    private final Semaphore blockBeforCreatingMarker = new Semaphore(1);
```
should be blockBefor**e**CreatingMarker.

steveloughran replied:
done. My inability to type well surfaces again

iwasakims commented:
LGTM overall. Tests against Tokyo region worked. I just left comments for nits.

```
$ mvn verify -Dtest=x -Dit.test=ITestS3AContractRename,ITestlRenameDeleteRace -Ds3guard
...
[INFO] Running org.apache.hadoop.fs.s3a.impl.ITestlRenameDeleteRace
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.044 s - in org.apache.hadoop.fs.s3a.impl.ITestlRenameDeleteRace
[INFO] Running org.apache.hadoop.fs.s3a.ITestS3GuardListConsistency
[INFO] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 35.603 s - in org.apache.hadoop.fs.s3a.ITestS3GuardListConsistency
[INFO] Running org.apache.hadoop.fs.contract.s3a.ITestS3AContractRename
[WARNING] Tests run: 22, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 12.523 s - in org.apache.hadoop.fs.contract.s3a.ITestS3AContractRename
```

iwasakims commented:
I got error on ITestS3AContractDistCp but this seems not to be related. I can not reproduce the error by running the ITestS3AContractDistCp alone by `-Dit.test=ITestS3AContractDistCp`.

```
$  mvn verify -Dtest=x -Dit.test=ITestS3AContract*
...
[INFO] Running org.apache.hadoop.fs.contract.s3a.ITestS3AContractDistCp
[ERROR] Tests run: 8, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 47.658 s <<< FAILURE! - in org.apache.hadoop.fs.contract.s3a.ITestS3AContractDistCp
[ERROR] testNonDirectWrite(org.apache.hadoop.fs.contract.s3a.ITestS3AContractDistCp)  Time elapsed: 4.055 s  <<< FAILURE!
java.lang.AssertionError: Expected 2 renames for a non-direct write distcp expected:<2> but was:<0>
        at org.apache.hadoop.fs.contract.s3a.ITestS3AContractDistCp.testNonDirectWrite(ITestS3AContractDistCp.java:98)
```

iwasakims commented:
> 3. don't downgrade FileNotFoundException to 'false' if source doesn't exist
> 4. raise FileAlreadyExistsException if dest path exists or parent is a file.

Should this JIRA be marked as incompatible for applicatinos assuming existing behavior?

steveloughran committed:
HADOOP-16721. Review integration test; make workflow clearer

Change-Id: I0dc69dd913d7f9ca36930ab09da1060bdfee8970

iwasakims commented on hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/impl/ITestRenameDeleteRace.java:
```diff
@@ -0,0 +1,248 @@
+import java.util.concurrent.TimeUnit;
+
+import com.amazonaws.AmazonClientException;
+import org.apache.hadoop.thirdparty.com.google.common.util.concurrent.ListeningExecutorService;
```
We need fix similar to #2758 after #2575.

steveloughran replied:
gotcha. Can we add some style checking to block this?

steveloughran replied:
no need, the compiler finds it. Will fix

steveloughran committed:
HADOOP-16271. Fix compilation, remove trailing whitespace from docs

Change-Id: I288e65899302adcebab5221508a005784cfe1d89

steveloughran commented:
testing: s3 london, most recently with `-Dparallel-tests -DtestsThreadCount=6 -Dmarkers=keep -Ds3guard -Ddynamo  -Dscale`

iwasakims reviewed (approved):
+1. The ITestRenameDeleteRace failed without the fix of S3AFileSystem as expected. The test passed with the fix regardless of `-Ds3guard`. Thanks, @steveloughran.

steveloughran commented:
ah, thanks -lovely. just tweaked the imports slightly for better backporting; I'll merge once the next compile is good, then do a cp and retest for branch-3.3.
